packages/ckeditor5-utils/src/dom/position.ts

```diff
@@ -99,12 +100,44 @@ export function getOptimalPosition( { element, target, positions, limiter, fitIn
 
 	let bestPosition: Position;
 
+	// Get intersection of all scrollable ancestors of `target`.
+	const allScrollableAncestors = getScrollableAncestors( target as HTMLElement );
+	const ancestorsIntersection: Rect | null = allScrollableAncestors
+		.reverse()
+		.map( item => new Rect( item as HTMLElement ) )
+		.reduce( ( ancestorIntersectionAccumulator, ancestorRect ): Rect => {
+			if ( !ancestorIntersectionAccumulator || ancestorIntersectionAccumulator.top === undefined ) {
+				return ancestorRect;
+			}
+
+			if ( !ancestorRect || ancestorRect.top === undefined ) {
+				return ancestorIntersectionAccumulator;
+			}
+
+			return ancestorIntersectionAccumulator.getIntersection( ancestorRect ) as Rect;
+		} );
+
 	// @if CK_DEBUG_POSITION // RectDrawer.clear();
 	// @if CK_DEBUG_POSITION // RectDrawer.draw( targetRect, { outlineWidth: '5px' }, 'Target' );
 
 	const viewportRect = fitInViewport && getConstrainedViewportRect( viewportOffsetConfig ) || null;
 	const positionOptions = { targetRect, elementRect, positionedElementAncestor, viewportRect };
 
+	if ( ( !ancestorsIntersection ) || ( !targetRect.getVisible() ) ) {
+		// Default last position is `viewportHidden`.
+		return new PositionObject( positions.at( -1 )!, positionOptions );
+	}
+	const ancestorsIntersectionWindowRect = new Rect( window ).getIntersection( ancestorsIntersection! );
```

Comment:
Get scrollable ancestors will always consider `global.document`. 

https://github.com/ckeditor/ckeditor5/blob/fece6640d3f587646caa42bae8d4dfa1dec6a742/packages/ckeditor5-utils/src/dom/getscrollableancestors.ts#L27

If you got it covered while retrieving rects, you wouldn't need to handle this one separately. It would be just another rect to intersect. See the implementation of the sticky panel to learn more

https://github.com/ckeditor/ckeditor5/blob/fece6640d3f587646caa42bae8d4dfa1dec6a742/packages/ckeditor5-ui/src/panel/sticky/stickypanelview.ts#L417-L429